### ISSUE TITLE
DLND-Week4(2)

### ISSUE THREAD

**Issue by chenghong-lin-nu** [Hclmaster/blog#11]
# Sequence to Sequence
## Intro to Seq to Seq
- Sequence to Sequence是一种特别的RNN结构。
- In this case, you **feed in a sequence of data** and the **network will output another sequence**.
- 就是说这里你输入进一序列的数据（比如说句子），然后网络会返回给你另一个数据（就相当于是翻译一样）。
- 会被应用在machine translation（机器翻译）等领域。
- 它的模型会类似于下面的 - many to many
![image](https://user-images.githubusercontent.com/16880879/37805751-1d3d2f1e-2e77-11e8-9309-5c6972b1c007.png)
- 但是最右边的有一个问题，就是我们希望RNN能够先把所有的inputs都读入进来再去产生输出；所以最右边的那个并不太适用。
- 左边那个可以map a sequence of any length to another sequence of any length.
- 原理：有两个RNN，一个读入数据，然后把它所学到的传递给下一个RNN，然后下一个就开始产生outputs了。
![image](https://user-images.githubusercontent.com/16880879/37805887-c8e5242a-2e77-11e8-8bbd-54f169597801.png)

## Applications
- Seq to Seq其实还是能够做很多事情的。
- English - Franch Translator
![image](https://user-images.githubusercontent.com/16880879/37806086-f063a192-2e78-11e8-969a-480c22198b48.png)
- Summarization Bot
![image](https://user-images.githubusercontent.com/16880879/37806099-0440947c-2e79-11e8-96a2-05865abae38e.png)
- Question - Answer Model: trained on a dataset and their answers
- ChatBot: trained on a dialog data
![image](https://user-images.githubusercontent.com/16880879/37806131-5856af4c-2e79-11e8-9162-eb26fec24efe.png)

## Architectures(架构)
- Overview:
![image](https://user-images.githubusercontent.com/16880879/37805887-c8e5242a-2e77-11e8-8bbd-54f169597801.png)
- 因为Encoder和Decoder都是RNNs，所以它们都有loops。
![image](https://user-images.githubusercontent.com/16880879/37809123-b8908540-2e88-11e8-9260-573457e8c846.png)
- Each time it reads an input, and do a transformation on its hidden state, and send that hidden state out to the next input step.
![image](https://user-images.githubusercontent.com/16880879/37809173-f43d465a-2e88-11e8-8ee5-f822cf66b6a6.png)
- 那么什么是**hidden state**呢？
- 最简单的情形下，就是a number of hidden units inside the cell.
- 但是在实际中，这大多数可能是hidden state inside the LSTM cell.
![image](https://user-images.githubusercontent.com/16880879/37809292-8f089734-2e89-11e8-8c1e-c6df736fa849.png)

## Architecture in more depth
![image](https://user-images.githubusercontent.com/16880879/37814529-754bc4da-2ea5-11e8-946d-cde8ac989725.png)
- 第一个参数是紫色的weights（input matrix of weights），我们每次take an input vector的时候都会乘以它。
- 然后我们有A参数（它是hidden state: travels between the time steps of the encoder and also out to the decoder），也就是说可以把现在拆分开来的看成是一个循环体。
- 然后还有另外一个weight matrix - v，然后就是乘以hidden_state，当decoder outputs a value的时候。
- 最后一个参数是B，这个参数就是在decoder循环的时候从一个时间段传到一个时间点。
- 一个十分重要的组成部分是fully connected layer -> translate the state of the network into an output we can choose.
- 然后这些呢，就是output logits，logits的数量和我们vocabulary的size是一样的。
![image](https://user-images.githubusercontent.com/16880879/37814947-b8078078-2ea7-11e8-9ead-0ad5768643b3.png)

## Preprocessing
- 用RNN构建的chat bot可以广泛地分为两个类别。
- **Retrieve-Based Model** 能够理解input text，能够从已经存在的回复当中选取最合适的回复。（相当于是从数据库中选取一个已经存在的并且是最合适的回复）。
- 它只适合于某一个特定场景，例如公司客服请求等，只要有够多的数据示例和归档数据来训练模型即可。
![image](https://user-images.githubusercontent.com/16880879/37815335-6d098b96-2ea9-11e8-8122-6ad91668bd9b.png)
- **Generative -based Models** 更难生成、而且技术还不算太成熟。
- 它是自己新生成一个回复、然后回复的质量呢，是依赖于我们数据集的质量。
![image](https://user-images.githubusercontent.com/16880879/37815471-eecca65e-2ea9-11e8-9ea5-fe800212d97c.png)
- 这里我们研究的是第二种。
- 我们需要处理我们的数据集看起来像下面那样：
![image](https://user-images.githubusercontent.com/16880879/37815499-115ad574-2eaa-11e8-9654-c8d277c7e7d1.png)
